src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -3542,24 +3566,85 @@ protected override void VisitArguments(ImmutableArray<BoundExpression> arguments
+            ImmutableArray<RefKind> refKindsOpt,
+            ImmutableArray<ParameterSymbol> parametersOpt,
+            ImmutableArray<int> argsToParamsOpt,
+            bool expanded)
         {
             Debug.Assert(!IsConditionalState);
             int n = arguments.Length;
-            if (n == 0)
+            if (n == 0 && parametersOpt.IsDefaultOrEmpty)
             {
-                return ImmutableArray<VisitArgumentResult>.Empty;
+                return (ImmutableArray<VisitArgumentResult>.Empty, argsToParamsOpt, refKindsOpt);
             }
-            var builder = ArrayBuilder<VisitArgumentResult>.GetInstance(n);
+
+            var visitedParameters = PooledHashSet<ParameterSymbol>.GetInstance();
+            var resultsBuilder = ArrayBuilder<VisitArgumentResult>.GetInstance(n);
             for (int i = 0; i < n; i++)
             {
-                FlowAnalysisAnnotations parameterAnnotations = GetCorrespondingParameter(i, parameters, argsToParamsOpt, expanded).Annotations;
-                builder.Add(VisitArgumentEvaluate(arguments[i], GetRefKind(refKindsOpt, i), parameterAnnotations));
+                var (parameter, _, parameterAnnotations, _) = GetCorrespondingParameter(i, parametersOpt, argsToParamsOpt, expanded);
+
+                resultsBuilder.Add(VisitArgumentEvaluate(arguments[i], GetRefKind(refKindsOpt, i), parameterAnnotations));
+                visitedParameters.Add(parameter);
+            }
+
+            if (!parametersOpt.IsDefaultOrEmpty)
+            {
+                ArrayBuilder<int>? argsToParamsBuilder;
+                if (argsToParamsOpt.IsDefault)
+                {
+                    argsToParamsBuilder = null;
+                }
+                else
+                {
+                    argsToParamsBuilder = ArrayBuilder<int>.GetInstance(parametersOpt.Length);
+                    argsToParamsBuilder.AddRange(argsToParamsOpt);
+                }
+
+                ArrayBuilder<RefKind>? argRefKindsBuilder;
+                if (refKindsOpt.IsDefault)
+                {
+                    argRefKindsBuilder = null;
+                }
+                else
+                {
+                    argRefKindsBuilder = ArrayBuilder<RefKind>.GetInstance(parametersOpt.Length);
+                    argRefKindsBuilder.AddRange(refKindsOpt);
+                }
+
+                var previousDisableNullabilityAnalysis = _disableNullabilityAnalysis;
+                _disableNullabilityAnalysis = true;
+                for (int i = 0; i < parametersOpt.Length; i++)
+                {
+                    var parameter = parametersOpt[i];
+                    // fill in unspecified optional arguments
+                    if (parameter.IsOptional && !visitedParameters.Contains(parameter))
+                    {
+                        var annotations = GetParameterAnnotations(parameter);
+                        if (!_defaultValues.TryGetValue((syntax, parameter), out var argument))
```

Comment:
>syntax [](start = 57, length = 6)

Why do we need the syntax as part of the key? I would think that ParameterSymbol -> BoundExpression would be a sufficient map, no?

Reply:
The LocalRewriter.GetDefaultParameterValue uses the provided Syntax to get the actual values when CallerMemberName, CallerLineNumber, etc. attributes are in use. The produced BoundExpression uses the containing invocation such as `BoundCall` for its `Syntax` property.

It's possible we could make it work without actually providing syntax to the BoundExpression for the default argument but I'm not sure what would break/need to change as a result.

Reply:
Makes sense. May be worth a comment

---
In reply to: [329248471](https://github.com/dotnet/roslyn/pull/38788#discussion_r329248471) [](ancestors = 329248471)

Reply:
Also, could you add a test with CallerExpression or some other attribute, if we don't have one already?

---
In reply to: [329250289](https://github.com/dotnet/roslyn/pull/38788#discussion_r329250289) [](ancestors = 329250289,329248471)

Reply:
Adding additional comments to the declaration of _defaultValues

---
In reply to: [329250289](https://github.com/dotnet/roslyn/pull/38788#discussion_r329250289) [](ancestors = 329250289,329248471)